### PR TITLE
Allow access to optimization direction in optimizer

### PR DESCRIPTION
**Pull request by moritzmeister**
Add reference in optimizer to the optimization direction. This is required for more advaned optimizers.

Also I am moving the setting of the references out of the init of the experimentdriver. To group things a bit more logically.

@ssheikholeslami can you double check, I am not breaking any ablation functionality. We should unify this across optimization and ablation, as we are passing these as arguments when instantiating the Ablator, whereas in optimization we set them separately.

### CODE REVIEW

File: maggy/core/experimentdriver.py [hunk 1]
```diff
@@ -243,6 +239,11 @@ def __init__(self, experiment_type, **kwargs):
         self.fd = hopshdfs.open_file(self.log_file, flags="w")
 
     def init(self, job_start):
+        # Set references to data in optimizer
+        self.optimizer.num_trials = self.num_trials
+        self.optimizer.searchspace = self.searchspace
+        self.optimizer.final_store = self._final_store
+        self.optimizer.direction = self.direction
```
Comment (ssheikholeslami) [inline maggy/core/experimentdriver.py]: Why not move this block to under L252?

Comment (ssheikholeslami) [reply maggy/core/experimentdriver.py]: No optimization object will be created if the experiment type is `ablation`, hence this will throw an exception.

**Review by ssheikholeslami — CHANGES_REQUESTED** [review]

File: maggy/core/experimentdriver.py (hunk 1 above)
Comment (ssheikholeslami) [reply maggy/core/experimentdriver.py]: To unify this across optimization and ablation, I propose something like `AbstractController(ABC)` that contains the common attributes and objects, and `AbstractOptimizer` and `AbstractAblator` would then inherit from it. We can address this in a later PR.

Comment (ssheikholeslami) [reply maggy/core/experimentdriver.py]: We can also implement a `ParameterServer` using an `AbstractController` - I have some ideas about this, let's discuss soon.

**Comment by moritzmeister** [comment]
Add also reference to trial-store in optimizer, same as for optimization direction

**Review by ssheikholeslami — APPROVED** [review]
LGTM

**Pull request merged**